File: src/Server/BookStore.Domain/Catalog/Models/Books/Genre.cs
namespace BookStore.Domain.Catalog.Models.Books;

using Common.Models;

public class Genre : Enumeration
{
    public static readonly Genre Horror = new(1, nameof(Horror));
    public static readonly Genre Fantasy = new(2, nameof(Fantasy));
    public static readonly Genre Mystery = new(3, nameof(Mystery));
    public static readonly Genre Romance = new(4, nameof(Romance));
    public static readonly Genre History = new(5, nameof(History));
    public static readonly Genre Biography = new(6, nameof(Biography));
    public static readonly Genre Fiction = new(7, nameof(Fiction));
    public static readonly Genre Drama = new(8, nameof(Drama));
    public static readonly Genre Dystopian = new(9, nameof(Dystopian));
    public static readonly Genre Adventure = new(10, nameof(Adventure));

    private Genre(int value)
        : this(value, FromValue<Genre>(value).Name)
    {
    }

    private Genre(int value, string name)
        : base(value, name)
    {
    }
}

File: src/Client/src/models/Genre.ts
import { Enumeration } from './common/Enumeration';

export class Genre extends Enumeration<Genre>() {
  static Horror = new Genre(1, 'Horror');
  static Fantasy = new Genre(2, 'Fantasy');
  static Mystery = new Genre(3, 'Mystery');
  static Romance = new Genre(4, 'Romance');
  static History = new Genre(5, 'History');
  static Biography = new Genre(6, 'Biography');
  static Fiction = new Genre(7, 'Fiction');
  static Drama = new Genre(8, 'Drama');
  static Dystopian = new Genre(9, 'Dystopian');

  static _ = Genre.closeEnum();

  name: string;
  label: string;
  value: number;

  constructor(value: number, name: string) {
    super();

    this.value = value;
    this.name = name;
    this.label = name;
  }
}